src/globus_sdk/services/compute/data.py

```diff
@@ -0,0 +1,47 @@
+import typing as t
+
+from globus_sdk import utils
+from globus_sdk._types import UUIDLike
+
+
+class FunctionMetadata(utils.PayloadWrapper):
+    """A wrapper for function metadata.
+
+    :param python_version: The Python version used to serialize the function.
+    :param sdk_version: The Globus Compute SDK version used to serialize the function.
+    """
+
+    def __init__(self, *, python_version: str, sdk_version: str):
+        super().__init__()
+        self["python_version"] = python_version
+        self["sdk_version"] = sdk_version
+
+
+class FunctionDocument(utils.PayloadWrapper):
+    """A function registration document.
+
+    :param function_name: The name of the function.
+    :param function_code: The serialized function source code.
+    :param description: The description of the function.
+    :param metadata: The metadata of the function.
+    :param group: Restrict function access to members of this Globus group.
+    :param public: Indicates whether the function is public.
+    """
+
+    def __init__(
+        self,
+        *,
+        function_name: str,
+        function_code: str,
+        description: t.Optional[str] = None,
+        metadata: t.Optional[FunctionMetadata] = None,
+        group: t.Optional[UUIDLike] = None,
+        public: bool = False,
```

Comment:
Should the defaults for these be that they are present with `null` values, or that they be omitted?

In newer document types and APIs we're trying to use `MISSING` instead of `None`. Very frequently, `None` has been used, historically, as a default where `MISSING` would be better. It doesn't apply to all APIs, however. `MISSING` isn't _inherently better_. It's just different in a way which we should consider.

For an example of usage, look at the GCS EndpointDocument type:
https://github.com/globus/globus-sdk-python/blob/6613e15a8928db39149b630441b630a73a1bcc23/src/globus_sdk/services/gcs/data/endpoint.py#L92

When the SDK is encoding a document as JSON to send, it _strips out_ `MISSING` values. So the effect is the difference between the following two encodings:
```python
# as written today
encode(
  FunctionDocument('foo', 'bar')
)
# {"function_name": "foo", "function_code": bar", "description": null, "metadata": null, "group": null, "public": False}

# if it were written using MISSING for defaults
encode(
  FunctionDocument('foo', 'bar')
)
# {"function_name": "foo", "function_code": bar"}
```

Note that I also assumed in this example that we make `public: bool | MissingType = MISSING` the default.

The difference comes into play most prominently with APIs which use a null field to indicate that a value should change / be removed in an update call. But it also matters here: if `public = False` is the default, then that's a client-side default. If `public = MISSING` is the default, then that's a server-side default.

---

If you're curious about how the behavior is implemented, here's MISSING:
https://github.com/globus/globus-sdk-python/blob/6613e15a8928db39149b630441b630a73a1bcc23/src/globus_sdk/utils.py#L25

And here's where MISSING gets filtered from payloads:
https://github.com/globus/globus-sdk-python/blob/6613e15a8928db39149b630441b630a73a1bcc23/src/globus_sdk/transport/encoders.py#L96-L99

(It also gets filtered from headers and query params.)

---

The exact behavior you will want depends on the behavior of the service, so I'm not advocating for a change so much as I'm making sure we discuss which is appropriate for the Compute API.

Reply:
The register function route currently handles null and missing values the same, so I'm fine with the proposed change.

Reply:
If they're equivalent (which is common for many APIs), I have a slight preference for MISSING. That way, if the service _adds_ a default behavior -- e.g., "default description is `<name> created by <user> on <date>`" -- the SDK picks it up seamlessly.